### PR TITLE
Modify Maya/3ds Max GLTF Standard Shader Spec/gloss to PBRMetalRoughness converter

### PR DESCRIPTION
**Pull request by Drigax**
Proposed fix for #466. This modifies the existing Standard Shader specular/gloss to Metal/Roughness converter to use the Babylon engine's specular power -> roughness conversion heuristic.

This change removes the baseColor specular color calculations, and removes the assumptions of surface metallic properties based of its shininess.

### CODE REVIEW

**Comment by deltakosh** [comment]
Pinging @bghgary for review

File: Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs [hunk 1]
```diff
@@ -562,47 +562,69 @@ private bool isTextureOk(BabylonTexture texture)
...
+        private MetallicRoughness ConvertToMetallicRoughness(SpecularGlossiness specularGlossiness, bool displayPrints = false)
+        {
+            // Hard coded points used to define the specular power to roughness curve.
+            var P0 = new BabylonVector2(0f, 1f);
+            var P1 = new BabylonVector2(0f, 0.1f);
+            var P2 = new BabylonVector2(0f, 0.1f);
+            var P3 = new BabylonVector2(1300f, 0.1f);
+
+            /**
+             * Helper function that defines the bezier curve as well.Given the control points, solve for x based on a given t for a cubic bezier curve
```
Comment (bghgary) [inline Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]:
```suggestion
             * Helper function that defines the bezier curve as well. Given the control points, solve for x based on a given t for a cubic bezier curve
```

File: Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs [hunk 2]
```diff
@@ -562,47 +562,69 @@ private bool isTextureOk(BabylonTexture texture)
...
-            return new MetallicRoughness
-            {
-                baseColor = baseColor,
-                opacity = opacity,
-                metallic = metallic,
-                roughness = 1 - glossiness
-            };
+        private MetallicRoughness ConvertToMetallicRoughness(SpecularGlossiness specularGlossiness, bool displayPrints = false)
+        {
+            // Hard coded points used to define the specular power to roughness curve.
```
Comment (bghgary) [inline Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: Are these points supposed to be identical to the glTF serializer in Babylon.js? Are the units for specular power the same between the standard material for Babylon and Max?

File: Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs [hunk 3]
```diff
@@ -562,47 +562,69 @@ private bool isTextureOk(BabylonTexture texture)
...
+            }
+
+            /*
+             * Helper function that calculates a roughness coefficient given a blinn-phong specular power coefficient
+             * @param specularPower the blinn-phong specular power coefficient
+             * @returns number result of specularPower -> roughness conversion curve.
+             */
+            float _solveForRoughness(float specularPower)
+            {
+                var t = Math.Pow(specularPower / P3.X, 0.333333);
```
Comment (bghgary) [inline Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: I can see this is copied from Babylon.js, but I can't remember why there is a power to the 1/3 here. @kcoley Do you remember?

File: Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs [hunk 4]
```diff
@@ -562,47 +562,69 @@ private bool isTextureOk(BabylonTexture texture)
...
+                return _cubicBezierCurve((float)t, P0.Y, P1.Y, P2.Y, P3.Y);
+            }
+
+            var diffuse = specularGlossiness.diffuse.scale(0.5f);
+            var opacity = specularGlossiness.opacity;
+            var glossiness = specularGlossiness.glossiness;
+            var metallic = 0;
+            var roughness = _solveForRoughness(glossiness * 256); // Glossiness = specularPower / 256
```
Comment (bghgary) [inline Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: Where does 256 come from?

Comment (Drigax) [reply Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: When we construct our SpecularGlossiness structure from the babylonMaterial we converted, we divide the specularPower by 256. I'm not entirely sure why...

https://github.com/BabylonJS/Exporters/blob/b45ec95831d630faff22f291192d4be542a11330/3ds%20Max/Max2Babylon/Exporter/BabylonExporter.GLTFExporter.Material.cs#L186

Comment (bghgary) [reply Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: But this is not a Babylon material right?

Comment (Drigax) [reply Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: and when we initially convert from Max->Babylon we, we multiply the shininess (Gloss in Max) by 256 to become what we define as SpecularPower in Babylon...strange...

Looking at the source, the shininess of a material is represented in max as a 0-1 value, while we represent it in Babylon as a 0-infinity value. I'm not sure why we use a factor of 256 other than to get this "looking right".

https://github.com/BabylonJS/Exporters/blob/b45ec95831d630faff22f291192d4be542a11330/3ds%20Max/Max2Babylon/Exporter/BabylonExporter.Material.cs#L152

File: Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs (hunk 2 above)
Comment (Drigax) [reply Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: yes, but not really?

here's two test renders I made using a directional light  and spheres of varying SpecularPower(Babylon)/glossiness(3dsMax) using the same values:

Babylon:
specspectum_babylon (3)

3dsMax:
specspectum_3dsmax

File: Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs (hunk 3 above)
Comment (Drigax) [reply Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: If we plot out our converter curve, we get this nice hyperbolic-ish curve:
shininess to roughness curve

File: Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs (hunk 2 above)
Comment (Drigax) [reply Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: It gets even worse when exporting from Max to Babylon, we get this:
screenshot_19-4-22_17-22


I'm willing to accept the disparity between different implementaitons of the Blinn lighting model as an unsolved mystery.

Comment (Drigax) [reply Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: as we discussed offline, this is caused by the 3dsMax->Babylon portion of the converter, I'm not quite sure if there is a linear relationship between 3dsMax and BabylonJS's shininess units.

But, I believe for the sake of "correctness" we can assume in our Babylon->GLTF part of the conversion, we are only working in Babylon units.

File: Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs (hunk 3 above)
Comment (Drigax) [reply Maya/Exporter/BabylonExporter.GLTFExporter.Material.cs]: I'll most likely do a deeper study in whether or not this is correct when I get around to tackling the 3dsMax->Babylon standard material conversion as well.

Hey Kacey, if you still have access to the data that defined this original curve, that would be a huge help.

**Pull request merged**